src/System.Net.Sockets/src/System/Net/Sockets/Socket.cs

```diff
@@ -4755,7 +4810,9 @@ public static bool ConnectAsync(SocketType socketType, ProtocolType protocolType
                 MultipleConnectAsync multipleConnectAsync = null;
                 if (dnsEP.AddressFamily == AddressFamily.Unspecified)
                 {
-                    multipleConnectAsync = new MultipleSocketMultipleConnectAsync(socketType, protocolType);
+                    multipleConnectAsync = SocketPal.SupportsMultipleConnectAttempts ?
+                        (MultipleConnectAsync)(new DualSocketMultipleConnectAsync(socketType, protocolType)) :
```

Comment:
Ah, actually `DualSocketMultipleConnectAsync` is a new class but GitHub diff is making a mess. 
I still don't understand why all these SocketPal.SupportsMultipleConnectAttempts exist: the Windows and even Linux code supports it. Why are we getting so much code churn to support OSX instead of pushing this code in PAL?

I think at this point we'd need some sort of architecture design that explains these changes. The CR became too complicated to understand the intent.